### PR TITLE
nixos/send: init

### LINKED ISSUE #350025: Package request: NixOS module `send`
**Project description**

Simple, private file sharing. A maintained fork of https://github.com/mozilla/send

The CLI-tool is already available in nixpkgs: https://search.nixos.org/packages?show=ffsend
I'm requesting a NixOS module / service.

**Metadata**

* homepage URL: https://send.vis.ee/
* source URL: https://gitlab.com/timvisee/send / https://github.com/timvisee/send
* license: MPL-2.0
* platforms: unix

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Moraxyc**
- **nixos/send: init**

must be merged after https://github.com/NixOS/nixpkgs/pull/350931

Closes https://github.com/NixOS/nixpkgs/issues/350025

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Moraxyc** [comment]
@74k1 please test this module

**Comment by Moraxyc** [comment]
@ofborg test send

File: nixos/modules/services/web-servers/send.nix [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    assertions = [
+      {
+        assertion = cfg.redis.createLocally -> cfg.redis.host == "localhost";
+        message = "the redis host must be localhost if services.send.redis.createLocally is set to true";
+      }
+    ];
+
+    networking.firewall.allowedTCPPorts = lib.optional cfg.openFirewall cfg.port;
+
+    users.users.send = lib.optionalAttrs (cfg.user == null) {
```
Comment (drupol) [inline nixos/modules/services/web-servers/send.nix:152]: Is it possible to use `dynamicUser` instead ?

File: pkgs/by-name/se/send/package.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
+{
+  lib,
+  buildNpmPackage,
+  fetchFromGitHub,
+  makeBinaryWrapper,
+  nodejs,
+}:
+
+buildNpmPackage rec {
```
Comment (drupol) [inline pkgs/by-name/se/send/package.nix:9]: you could add `nixosTests` in this file.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi,

Just made a quick review.

File: nixos/modules/services/web-servers/send.nix (hunk 1 above)
Comment (Moraxyc) [reply nixos/modules/services/web-servers/send.nix]: I'm not sure whether users should be given the right to freely choose `dataDir`. 
Adding DynamicUser means that DataDir must be StateDirectory.

Comment (drupol) [reply nixos/modules/services/web-servers/send.nix]: Have you checked some other services for inspiration ?

Comment (Moraxyc) [reply nixos/modules/services/web-servers/send.nix]: Yeah, I have tried to find methods in other modules, and the most common one is what I am currently using. There may also be solutions that I have not seen. 

The most important problem is that after enabling `DynamicUser`, it is impossible to grant the read/write or ownership of the path to this user. Even using ReadWritePaths only solves the file system access problem in the sandbox, not the ownership/read/write problem.

Comment (Moraxyc) [reply nixos/modules/services/web-servers/send.nix]: `ExecStartPre=+` may be a solution

**Comment by NiklasGollenstede** [comment]
Can't say I had heard about this `send` before, but the module looks nice!

systemd hardening, DynamicUser, systemd-credentials, not inherently any secrets in the store :+1:

Is just "send" an appropriate package name, tough? Do other distributions also simply call it "send"?

**Review by drupol — APPROVED** [review]

File: nixos/modules/services/web-servers/send.nix (hunk 1 above)
Comment (Moraxyc) [reply nixos/modules/services/web-servers/send.nix]: Considering the stateless and expired nature of the files in Send, IMO it is OK to directly set `dataDir` to stateDirectory. The original `config.js` creates a random folder in /tmp.

**Comment by drupol** [comment]
Can you rebase this PR now that https://github.com/NixOS/nixpkgs/pull/350931 has been merged?

**Comment by Moraxyc** [comment]
> Can't say I had heard about this `send` before, but the module looks nice!
> 
> systemd hardening, DynamicUser, systemd-credentials, not inherently any secrets in the store 👍
> 
> Is just "send" an appropriate package name, tough? Do other distributions also simply call it "send"?

I have searched the entire `nixpkgs` and found no conflicting modules or package names.
I agree with you that `Send` sounds a bit too common, but it is indeed the name upstream

**Comment by Moraxyc** [comment]
> Can you rebase this PR now that #350931 has been merged?

Done.

**Review by drupol — APPROVED** [review]

**Comment by 74k1** [comment]
Awesome PR! looking forward to it :)

File: nixos/modules/services/web-servers/send.nix [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+          with types;
+          attrsOf (
+            nullOr (oneOf [
+              bool
+              int
+              str
+              (listOf int)
+            ])
+          );
+        description = ''
```
Comment (dotlambda) [inline nixos/modules/services/web-servers/send.nix:29]:
```suggestion
        default = { };
        description = ''
```

File: nixos/modules/services/web-servers/send.nix [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        message = "the redis host must be localhost if services.send.redis.createLocally is set to true";
+      }
+    ];
+
+    networking.firewall.allowedTCPPorts = lib.optional cfg.openFirewall cfg.port;
+
+    services.redis = lib.optionalAttrs cfg.redis.createLocally {
+      servers."${cfg.redis.name}" = {
+        enable = true;
+        bind = "localhost";
```
Comment (dotlambda) [inline nixos/modules/services/web-servers/send.nix:147]: Please use a Unix socket by default.

**Pull request merged**